Sources/SwiftFormat/API/Selection.swift

```diff
@@ -0,0 +1,76 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import SwiftSyntax
+
+/// The selection as given on the command line - an array of offets and lengths
+public struct Selection {
+  // TODO: make this a typealias instead
+  public struct Range: Decodable {
+    public let offset: AbsolutePosition
+    public let length: Int
+    public var end: AbsolutePosition { offset.advanced(by: length) }
+
+    public init(from decoder: any Decoder) throws {
+      let container = try decoder.container(keyedBy: CodingKeys.self)
+      offset = AbsolutePosition(utf8Offset: try container.decode(Int.self, forKey: .offset))
+      length = try container.decode(Int.self, forKey: .length)
+    }
+
+    public init(start: AbsolutePosition, end: AbsolutePosition) {
+      offset = start
+      length = end.utf8Offset - start.utf8Offset
+    }
+
+    enum CodingKeys: CodingKey {
+      case offset
+      case length
+    }
+  }
+
+  public let ranges: [Range]
+
+  public init?(json: String) {
+    guard let data = json.data(using: .utf8),
+          let ranges = try? JSONDecoder().decode([Range].self, from: data) else {
+      return nil
+    }
+    self.ranges = ranges
+  }
+
+  public func contains(_ position: AbsolutePosition) -> Bool {
+    return ranges.contains { return $0.offset <= position && position < $0.end }
+  }
+
+  public func overlaps(_ range: Range) -> Bool {
+    return ranges.contains {
+      let maxStart = max($0.offset, range.offset)
+      let minEnd = (min($0.end, range.end))
+      return maxStart < minEnd
+    }
+  }
+}
+
+
+public extension Syntax {
+  /// return true if the node is completely inside any range in the selection (or if there's
+  /// no selection)
+  func isInsideSelection(_ selection: Selection?) -> Bool {
```

Comment:
What if we had a default-initialized `Selection` represent an "infinite" selection that contains any offset? Then we could get rid of the optionality everywhere, simplifying the API.

Then, with a `Selection` always being present, we could invert this API and make it a method on `Selection` (`Selection.contains(_ node: Syntax)`) instead of a `Syntax` extension.

Reply:
Actually, one thing I want to do is disable the handling of `swift-format-ignore` comments when there's a selection. (If a user's making a selection, then they might actually want to fix part of the code where formatting is normally disabled.) This is one case where we actually need to know if there's a selection. There might end up being others. Keeping `selection` optional allows for this.

(I'm not at all averse to inverting that API though. I don't like that it currently works in both directions. Should be cleaner.)

Reply:
I think I disagree with that; a `swift-format-ignore` comment is an explicit request that the formatter not touch that section of code under any circumstances.

Imagine this workflow: a user has some nested array/dictionary structure that they've meticulously aligned into columns, so they use `swift-format-ignore` to ensure that their horizontal spaces don't get collapsed. Then, they make some changes to that table (add a new row, or move things around), and finally, they run a hypothetical `git-swift-format` (à la `git-clang-format`) that passes a selection representing the lines that they changed. With the proposed behavior, it would reformat parts of that table that they asked not to be changed.

If you want to provide a way to "ignore the `ignore`s", I think that should be a separate feature (I'm not sure how useful it would be) and is orthogonal to whether a selection is passed.